File: flutter_app/test/features/quoting/domain/usecase/calculate_quote_use_case_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_app/features/quoting/domain/model/quote_models.dart';
import 'package:flutter_app/features/quoting/domain/repository/tariff_repository.dart';
import 'package:flutter_app/features/quoting/domain/usecase/calculate_quote_use_case.dart';

class FakeTariffRepository implements TariffRepository {
  double multiplier = 1.0;
  @override
  Future<double> getRemoteMultiplier(String zipCode) async => multiplier;
}

void main() {
  late CalculateQuoteUseCase useCase;
  late FakeTariffRepository repository;

  setUp(() {
    repository = FakeTariffRepository();
    useCase = CalculateQuoteUseCase(repository);
  });

  group('CalculateQuoteUseCase (Dart) Tests', () {
    test('should return validation error for weight <= 0', () async {
      final request = QuoteRequest(
        weightKg: 0,
        distanceKm: 100,
        shippingType: ShippingType.standard,
        destinationZipCode: '12345',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteError>());
      expect((result as QuoteError).code, 'INVALID_WEIGHT');
    });

    test('should apply 25% surcharge for foreign zip codes (01-05)', () async {
      // Base: 50 + (8*10) + (2*100) = 330
      // Foreign: 330 * 1.25 = 412.5
      final request = QuoteRequest(
        weightKg: 10,
        distanceKm: 100,
        shippingType: ShippingType.standard,
        destinationZipCode: '01000',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteSuccess>());
      expect((result as QuoteSuccess).data.finalPrice, 412.5);
    });

    test('should apply special handling surcharge for weight > 20kg', () async {
      // Base: 50 + (8*25) + (2*100) = 450
      // Special: 450 + 100 = 550
      final request = QuoteRequest(
        weightKg: 25,
        distanceKm: 100,
        shippingType: ShippingType.standard,
        destinationZipCode: '99999',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteSuccess>());
      expect((result as QuoteSuccess).data.finalPrice, 550.0);
      expect((result as QuoteSuccess).data.details.specialHandlingApplied, isTrue);
    });

    test('exactly 20kg should NOT trigger special handling', () async {
      final request = QuoteRequest(
        weightKg: 20,
        distanceKm: 100,
        shippingType: ShippingType.standard,
        destinationZipCode: '99999',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteSuccess>());
      expect((result as QuoteSuccess).data.details.specialHandlingApplied, isFalse);
    });

    test('exactly 200km should result in 2 days standard', () async {
      final request = QuoteRequest(
        weightKg: 10,
        distanceKm: 200,
        shippingType: ShippingType.standard,
        destinationZipCode: '99999',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteSuccess>());
      expect((result as QuoteSuccess).data.estimatedDays, 2);
    });

    test('200.1km should trigger 3rd day due to rounding up', () async {
      final request = QuoteRequest(
        weightKg: 10,
        distanceKm: 200.1,
        shippingType: ShippingType.standard,
        destinationZipCode: '99999',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteSuccess>());
      expect((result as QuoteSuccess).data.estimatedDays, 3);
    });

    test('express delivery should round up half days', () async {
      // 200.1km -> 3 days standard. Express: 3 / 2 = 1.5 -> ceil = 2
      final request = QuoteRequest(
        weightKg: 10,
        distanceKm: 200.1,
        shippingType: ShippingType.express,
        destinationZipCode: '99999',
      );

      final result = await useCase(request);

      expect(result, isA<QuoteSuccess>());
      expect((result as QuoteSuccess).data.estimatedDays, 2);
    });
  });
}

File: core/src/commonTest/kotlin/com/jght/sjrqromx/business/shipping/kmp_shipping_platform/features/quoting/domain/usecase/CalculateQuoteUseCaseTest.kt
package com.jght.sjrqromx.business.shipping.kmp_shipping_platform.features.quoting.domain.usecase

import com.jght.sjrqromx.business.shipping.kmp_shipping_platform.features.quoting.domain.model.*
import com.jght.sjrqromx.business.shipping.kmp_shipping_platform.features.quoting.domain.repository.TariffRemoteService
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class FakeTariffRemoteService(
    var multiplier: Double = 1.0,
    var shouldFail: Boolean = false
) : TariffRemoteService {
    override suspend fun getRemoteMultiplier(zipCode: String): Result<Double> {
        if (shouldFail) return Result.failure(Exception("Service Failure"))
        return Result.success(multiplier)
    }
}

class CalculateQuoteUseCaseTest {

    private val fakeService = FakeTariffRemoteService()
    private val useCase = CalculateQuoteUseCase(fakeService)

    @Test
    fun `when weight is zero then return validation error`() = runTest {
        val request = QuoteRequest(0.0, 100.0, ShippingType.STANDARD, "01234")
        val result = useCase(request)
        
        assertTrue(result is QuoteResult.Error)
        assertEquals(QuoteErrorType.VALIDATION_ERROR, (result as QuoteResult.Error).error.type)
        assertEquals("INVALID_WEIGHT", result.error.code)
    }

    @Test
    fun `when distance is zero then return validation error`() = runTest {
        val request = QuoteRequest(10.0, 0.0, ShippingType.STANDARD, "01234")
        val result = useCase(request)
        
        assertTrue(result is QuoteResult.Error)
        assertEquals(QuoteErrorType.VALIDATION_ERROR, (result as QuoteResult.Error).error.type)
        assertEquals("INVALID_DISTANCE", result.error.code)
    }

    @Test
    fun `calculate standard quote correctly`() = runTest {
        // Base: 50 + (8*10) + (2*100) = 50 + 80 + 200 = 330
        // Foreign: 330 * 1.25 = 412.5
        // Remote: 412.5 * 1.0 = 412.5
        fakeService.multiplier = 1.0
        val request = QuoteRequest(10.0, 100.0, ShippingType.STANDARD, "01234")
        val result = useCase(request)
        
        assertTrue(result is QuoteResult.Success)
        assertEquals(412.5, (result as QuoteResult.Success).data.finalPrice)
        assertEquals(2, result.data.estimatedDays) // 1 + ceil(100/200) = 1 + 1 = 2
    }

    @Test
    fun `calculate express quote correctly`() = runTest {
        // Base: 50 + (8*10) + (2*100) = 330
        // Foreign: 330 * 1.25 = 412.5
        // Express: 412.5 * 1.40 = 577.5
        // Remote: 577.5 * 1.0 = 577.5
        fakeService.multiplier = 1.0
        val request = QuoteRequest(10.0, 100.0, ShippingType.EXPRESS, "01234")
        val result = useCase(request)
        
        assertTrue(result is QuoteResult.Success)
        assertEquals(577.5, (result as QuoteResult.Success).data.finalPrice)
        assertEquals(1, result.data.estimatedDays) // ceil(2/2) = 1
    }

    @Test
    fun `apply special handling when weight is over 20kg`() = runTest {
        // Base: 50 + (8*25) + (2*100) = 50 + 200 + 200 = 450
        // Special Handling: 450 + 100 = 550
        // Standard Zip (starts with 9): 550
        // Remote: 550 * 1.1 = 605.0
        fakeService.multiplier = 1.1
        val request = QuoteRequest(25.0, 100.0, ShippingType.STANDARD, "99999")
        val result = useCase(request)
        
        assertTrue(result is QuoteResult.Success)
        assertEquals(605.0, (result as QuoteResult.Success).data.finalPrice)
        assertTrue(result.data.details.specialHandlingApplied)
    }

    @Test
    fun `return error when remote service fails`() = runTest {
        fakeService.shouldFail = true
        val request = QuoteRequest(10.0, 100.0, ShippingType.STANDARD, "01234")
        val result = useCase(request)
        
        assertTrue(result is QuoteResult.Error)
        assertEquals(QuoteErrorType.REMOTE_SERVICE_ERROR, (result as QuoteResult.Error).error.type)
    }
}